enhancements/mcd-reboot-dictionary.md

```diff
@@ -0,0 +1,275 @@
+---
+title: mcd-reboot-dictionary
+authors:
+  - "@beekhof"
+reviewers:
+  - "@crawford"
+  - "@eparis"
+  - "@rphillips"
+  - "@runcom"
+  - "@yuqi-zhang"
+  - "@kikisdeliveryservice"
+  - "@sinnykumari"
+  - "@ericavonb"
+approvers:
+  - "@crawford"
+  - "@eparis"
+  - "@rphillips"
+  - "@runcom"
+creation-date: 2019-12-15
+last-updated: 2019-12-15
+status: provisional
+see-also:
+  - NA
+replaces:
+  - NA
+superseded-by:
+  - NA
+---
+
+# MCD Reboot Dictionary
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+- Should we allow the whitelist to apply recursively to a directory, or make
+  use of pattern matching?
+
+- Is there any way to mitigate the version skew problem?
+
+- Should the parts of the configuration that do and don't require reboot be
+  part of user facing documentation?
+
+
+## Summary
+
+A node's configuration is managed by the MCO.  When the configuation is changed
+by the system or an admin, the MCO and MCD write a new version of the file to
+disk and reboot the system to ensure it is applied.
+
+This proposal is to create a dictionary that defines a whitelist of file paths
+for which alternative application strategies should be used.  
+
+
+## Motivation
+
+Always rebooting a node is reasonable for cloud based systems, since they can
+be rebooted very quickly.  Additionally, it is the only way for kernel
+parameter changes to take effect.
+
+However on baremetal systems, rebooting a node takes significantly longer (in
+the order of minutes) and the machine configuration includes files that need
+either no action to take effect, or can be more quickly applied by restarting a
+systemd service.
+
+Telco admins therefor want the system to use the least invasive method for 
+applying configuration file updates, to reduce periods of degraded system and 
+application availability.
+
+### Goals
+
+- Creation of a minimally invasive solution that can be delivered in the 
+  short-term, allowing time for a more comprehensive approach to be developed
+- A system defined whitelist of file paths that do not trigger a reboot
```

Comment:
should we be thinking of this in terms of white-listing paths or supporting optimized reboots for certain functions? im not convinced that whitelisting paths won't cause some other issues and is overbroad, whereas iterating on optimizing specific reboots for specific functionalities ie "don't reboot a node when you update ssh keys" is more scalpel-like with more predictable results and a clear articulation of what is supported and what is not.

Reply:
I hear what you're saying about focusing on specific functions, but don't they all eventually decompose into "i wrote this file(s) to disk, what do I do about it?"

Surely something, somewhere, needs know which file(s) a given CRD field is associated with, and what needs to be done to apply it.

Is there a practical difference between a lookup table that only has an entry for `${XYZ}` (authkeys for the least complex example), and adding conditionals through the code that check if the current file is `${XYZ}`?  Or is the thinking that responsibility for maintaining `${XYZ}` would be transferred to another operator?

Reply:
An interesting perspective @beekhof but one I'm not sold on. Adding a file to /etc/containers/registry.d/ is something I could agree is safe to do without a reboot. Removing such a file I'd argue is not. Thus you can't say /etc/container/registry.d/* can be manipulated without reboot.

It takes domain specific knowledge. And it's really tricky because I'm arguing for 'immediate consistency' in an 'eventual consistency' system. I know I'm arguing out of 2 sides of my mouth, but apply a change that won't cause problems until some time in the future is much worse than a reboot. Weekly reboots are normal. We design for that to be the norm.